topdown/eval.go

```diff
@@ -2892,6 +2910,100 @@ func (e evalTerm) save(iter unifyIterator) error {
 
 }
 
+type evalEvery struct {
+	e         *eval
+	expr      *ast.Expr
+	generator ast.Body
+	body      ast.Body
+}
+
+func (e evalEvery) eval(iter unifyIterator) error {
+	// unknowns in domain or body: save the expression, PE its body
+	// NOTE(sr): if the domain is not unknown, we could unroll the expression
+	if e.e.unknown(e.generator, e.e.bindings) || e.e.unknown(e.body, e.e.bindings) {
+		return e.save(iter)
+	}
+
+	child := e.e.closure(e.generator)
+	all := true // all generator evaluations yield one successful body evaluation
+	err := child.Run(func(child *eval) error {
+		if !all {
+			// NOTE(sr): Is this good enough? We don't have a "fail EE".
+			// This would do extra work, like iterating needlessly if domain was a large array.
+			return nil
+		}
+		body := child.closure(e.body)
+		body.findOne = true
+		done := false
+		err := body.eval(func(*eval) error {
+			done = true
+			return nil
+		})
+		if !done {
+			all = false
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	if all {
+		return iter()
+	}
+	return nil
+}
+
+func (e *evalEvery) save(iter unifyIterator) error {
+	cpy := e.expr.Copy()
+	every := cpy.Terms.(*ast.Every)
+
+	child := e.e.child(e.body)
+	child.traceEnter(e.expr) // ?
+
+	var savedQueries ast.Body
+	e.e.saveStack.PushQuery(nil)
+
+	kvVars := every.KeyValueVars()
+	vars := make([]*ast.Term, 0, len(kvVars))
+	for _, v := range kvVars.Sorted() {
+		vars = append(vars, ast.NewTerm(v))
+	}
+	e.e.saveSet.Push(vars, child.bindings)
+
+	err := child.eval(func(child *eval) error {
+		child.traceExit(e.expr)
+
+		query := e.e.saveStack.Peek()
+		plugged := query.Plug(child.bindings) // TODO(sr): correct bindings?
```

Comment:
NB I guess to much when it comes to bindings. What's the right mental model here? 💭 e.e.bindings, e.e.caller.bindings, child.bindings, pick two of them for PlugNamespaced... 😮